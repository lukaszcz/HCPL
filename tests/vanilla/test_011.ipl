
// higher-order recursive functions test

let f0 = \&f \&g \&x {
   print 6;
   print x;
   print 7;
   if (not (= x 0)) then {
      print 9;
      f f g (g x)
   } else
      ()
};

let g = \&x {
   print (* 10 x);
   - x 1
};

let f1 = \&f \&g \!x {
   print 6;
   print x;
   print 7;
   if (not (= x 0)) then {
      print 9;
      f f g (g x)
   } else
      ()
};

let f = f0 f0 g;
let fe = f0 f1 g;

print (f 4);
print (fe 4);
