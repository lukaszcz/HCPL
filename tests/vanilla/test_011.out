6
4
7
9
6
40
3
7
9
6
30
2
7
9
6
20
1
7
9
6
10
0
7
()
6
4
7
9
40
6
3
7
9
30
6
2
7
9
20
6
1
7
9
10
6
0
7
()
